api/src/modules/impact/impact.service.ts

```diff
@@ -0,0 +1,169 @@
+  growthRate: number = 1.5;
+  logger: Logger = new Logger(ImpactService.name);
+
+  constructor(
+    public readonly indicatorService: IndicatorsService,
+    public readonly sourcingRecordService: SourcingRecordsService,
+    public readonly sourcingLocationService: SourcingLocationsService,
+  ) {}
+  async getImpactTable(
+    impactTableDto: GetImpactTableDto,
+  ): Promise<ImpactTable> {
+    const indicators: Indicator[] =
+      await this.indicatorService.getIndicatorsById(
+        impactTableDto.indicatorIds,
+      );
+    this.logger.log('Retrieving data from DB to build Impact Table...');
+
+    const dataForImpactTable: ImpactTableData[] =
+      await this.sourcingRecordService.getDataForImpactTable(impactTableDto);
+    return this.buildImpactTable(
+      impactTableDto,
+      indicators,
+      dataForImpactTable,
+    );
+  }
+
+  private buildImpactTable(
+    queryDto: GetImpactTableDto,
+    indicators: Indicator[],
+    dataForImpactTable: ImpactTableData[],
+  ): ImpactTable {
+    this.logger.log('Building Impact Table...');
+    const { groupBy, startYear, endYear } = queryDto;
+    const impactTable: ImpactTableDataByIndicator[] = [];
+    const rangeOfYears: number[] = range(startYear, endYear + 1);
+    indicators.forEach((indicator: Indicator, i: number) => {
+      impactTable.push({
+        indicatorShortName: indicator.shortName as string,
+        indicatorId: indicator.id,
+        groupBy: groupBy,
+        rows: [],
+        yearSum: [],
+        metadata: { unit: indicator.unit.symbol },
+      });
+      const dataByIndicator: ImpactTableData[] = dataForImpactTable.filter(
+        (data: ImpactTableData) => data.indicatorId === indicator.id,
+      );
+      const namesByIndicator: string[] = [
+        ...new Set(dataByIndicator.map((el: ImpactTableData) => el.name)),
+      ];
+
+      for (const [index, name] of namesByIndicator.entries()) {
+        impactTable[i].rows.push({ name: name, values: [] });
+        let rowIndex: number = 0;
+        for (const year of rangeOfYears) {
+          const dataForYear: ImpactTableData | undefined = dataByIndicator.find(
+            (data: ImpactTableData) => {
+              return (data.year as unknown as number) === year;
+            },
+          );
+          if (dataForYear) {
+            impactTable[i].rows[index].values.push({
+              year: parseInt(dataForYear.year),
+              value: dataForYear.impact,
+              isProjected: false,
+            });
+            ++rowIndex;
+          } else {
+            const lastYearsValue: number =
+              impactTable[i].rows[index].values[rowIndex - 1].value;
+            impactTable[i].rows[index].values.push({
+              year: year,
+              value: lastYearsValue + (lastYearsValue * this.growthRate) / 100,
+              isProjected: true,
+            });
+            ++rowIndex;
```

Comment:
It can not be moved out, as that index need to increase by one everytime a new year either projected or not is saved, so the next one can get the last value, being projected or not

Reply:
not clear why this cannot be moved. again: you have it inside the `if` and the `else`. not sure why it can't be just after the else closes...